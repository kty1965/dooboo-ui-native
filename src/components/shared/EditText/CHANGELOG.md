[0.4.1]
- Support row aligned.

[0.3.0]
- First release.
