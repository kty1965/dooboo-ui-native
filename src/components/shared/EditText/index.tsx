import {
  NativeSyntheticEvent,
  Platform,
  TextInputProps,
  TextInputSubmitEditingEventData,
  TextStyle,
  ViewStyle,
} from 'react-native';
import React, { ReactElement, useState } from 'react';

import styled from 'styled-components/native';

const Container = styled.View`
  flex-direction: column;
  align-self: stretch;
  width: 100%;
`;

const UnderLine = styled.View`
  border: 0.6px solid #eaeaf9;
`;

const Title = styled.Text`
  font-size: 14px;
  font-weight: 500;
  color: #b9b9c4;
`;

const StyledTextInput = Platform.select({
  ios: styled.TextInput`
    padding: 15px 0 10px 0;
    font-size: 15px;
    font-weight: 500;
    color: #2c374e;
  `,
  android: styled.TextInput`
    padding-left: 0px;
    padding-bottom: 5px;
    font-size: 15px;
    font-weight: 500;
    color: #2c374e;
  `,
});

const StyledInvalidText = styled.Text`
  font-size: 12px;
  font-weight: 500;
  margin-top: 5px;
  color: #ff8989;
`;

interface Props {
  testID?: string;
  errorTestID?: string;
  style?: ViewStyle;
  underlineStyle?: ViewStyle;
  label?: string;
  textStyle?: TextStyle;
  labelTextStyle?: TextStyle;
  errorTextStyle?: TextStyle;
  errorText?: string;
  value?: TextInputProps['value'];
  placeholder?: TextInputProps['placeholder'];
  placeholderTextColor?: TextInputProps['placeholderTextColor'];
  secureTextEntry?: TextInputProps['secureTextEntry'];
  onChangeText?: TextInputProps['onChangeText'];
  onSubmitEditing?: (
    e: NativeSyntheticEvent<TextInputSubmitEditingEventData>,
  ) => void;
  focusColor?: string;
  errorColor?: string;
  autoCapitalize?: TextInputProps['autoCapitalize'];
  textInputProps?: TextInputProps;
}

function EditText(props: Props): ReactElement {
  const [focused, setFocus] = useState(false);

  const {
    testID,
    errorTestID,
    style,
    underlineStyle,
    label,
    textStyle,
    labelTextStyle,
    errorTextStyle,
    errorText,
    value,
    placeholder,
    placeholderTextColor,
    secureTextEntry,
    onChangeText,
    onSubmitEditing,
    focusColor,
    errorColor,
    autoCapitalize = 'none',
    textInputProps,
  } = props;

  return (
    <Container style={style}>
      <Title
        style={[
          // prettier-ignore
          focused
            ? { color: focusColor }
            : errorText
              ? { color: errorColor }
              : null,
          labelTextStyle,
        ]}
      >
        {label}
      </Title>
      <StyledTextInput
        {...textInputProps}
        testID={testID}
        autoCapitalize={autoCapitalize}
        onFocus={(): void => setFocus(true)}
        onBlur={(): void => setFocus(false)}
        placeholder={placeholder}
        placeholderTextColor={placeholderTextColor}
        value={value}
        style={textStyle}
        onChangeText={onChangeText}
        secureTextEntry={secureTextEntry}
        onSubmitEditing={onSubmitEditing}
      />
      <UnderLine
        style={[
          // prettier-ignore
          focused
            ? { borderColor: focusColor }
            : errorText
              ? { borderColor: errorColor }
              : null,
          underlineStyle,
        ]}
      />
      {errorText ? (
        <StyledInvalidText
          testID={errorTestID}
          style={[
            {
              color: errorColor,
            },
            errorTextStyle,
          ]}
        >
          {` ${errorText} `}
        </StyledInvalidText>
      ) : null}
    </Container>
  );
}

EditText.defaultProps = {
  focusColor: '#79B3F5',
  errorColor: '#FF8989',
};

export default EditText;
