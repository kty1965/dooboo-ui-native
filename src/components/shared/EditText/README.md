# EditText

[![Npm Version](http://img.shields.io/npm/v/@dooboo-ui/native-edit-text.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-edit-text)
[![Downloads](http://img.shields.io/npm/dm/@dooboo-ui/native-edit-text.svg?style=flat-square)](https://npmjs.org/package/@dooboo-ui/native-edit-text)

> [EditText] component is an enhanced version on pure react-native [TextInput] component.

![image](https://user-images.githubusercontent.com/27461460/62299053-28e6c680-b4af-11e9-8c59-61b79ad82c4d.png)

## Props

|                      | necessary | types                  | default |
| -------------------- | --------- | ---------------------- | ------- |
| testID               |           | string                 |         |
| errorTestID          |           | string                 |         |
| isRow                |           | boolean                |         |
| style                |           | `StyleProp<ViewStyle>` |         |
| underlineStyle       |           | `StyleProp<ViewStyle>` |         |
| label                |           | string                 |         |
| textStyle            |           | `StyleProp<TextStyle>` |         |
| labelTextStyle       |           | `StyleProp<TextStyle>` |         |
| errorTextStyle       |           | `StyleProp<TextStyle>` |         |
| errorText            |           | string                 |         |
| value                |           | string                 |         |
| placeholder          |           | string                 |         |
| placeholderTextColor |           | string                 |         |
| secureTextEntry      |           | boolean                |         |
| onSubmitEditing      |           | func                   |         |
| onChangeText         |           | (e) => {}              |         |
| textInpuProps        |           | `TextInputProps`       |         |

## Installation

```sh
yarn add @dooboo-ui/native
```

or

```sh
yarn add @dooboo-ui/native-edit-text
```

## Getting started

- Import

  ```javascript
  import { EditText } from '@dooboo-ui/native';
  // or
  import EditText from '@dooboo-ui/native-edit-text';
  ```

- Usage

  ```javascript
  function Page(props: Props) {
    const validateEmail = (email: string) => {
      const re = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
      return re.test(email);
    };

    const fontStyle: TextStyle = {
      fontWeight: 'bold',
      fontSize: 13,
    };

    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [errorEmail, setErrorEmail] = useState('');

    const onSignIn = () => {
      if (!validateEmail(email)) {
        setErrorEmail('Not a valid email address');
      } else {
        setErrorEmail('');
      }
    };

    const onTextChanged = (type: string, text: string) => {
      type === 'EMAIL' ? setEmail(text) : setPassword(text);

      if (type === 'EMAIL' && text === '') {
        setErrorEmail('');
      }
    };

    return (
      <StyledScrollView
        contentContainerStyle={{
          marginTop: 8,
          paddingHorizontal: 20,
          paddingBottom: 40,
        }}
      >
        <Container>
          <HeaderTitle>Sign in with Email</HeaderTitle>
          <EditText
            testID="EMAIL_INPUT"
            textStyle={{
              color: '#495057',
            }}
            label="Email"
            placeholder="Write email address"
            placeholderTextColor="#ADB5BD"
            text={email}
            onTextChanged={(text: string) => onTextChanged('EMAIL', text)}
            style={{ marginTop: 50 }}
            errorText={errorEmail}
            onSubmitEditing={onSignIn}
          />
          <EditText
            testID="PASSWORD_INPUT"
            textStyle={{
              color: '#ADB5BD',
            }}
            secureTextEntry={true}
            label="Password"
            placeholder="Please write your password"
            placeholderTextColor="#ADB5BD"
            text={password}
            onTextChanged={(text: string) => onTextChanged('PASSWORD', text)}
            style={{ marginTop: 36 }}
            onSubmitEditing={onSignIn}
          />
          <StyledSignInButton
            testID="btnEmail"
            onClick={() => onSignIn()}
            textStyle={fontStyle}
            text="Login"
          />
          {/* Email SignUp text */}
          <View
            style={{
              marginTop: 20,
              flexDirection: 'row',
              justifyContent: 'center',
              alignItems: 'center',
            }}
          >
            <StyledText testID="NO_ACCOUNT">
              Do not have and account?{' '}
            </StyledText>
            <TouchableOpacity onPress={() => null} style={{ padding: 4 }}>
              <StyledAccentText>Find</StyledAccentText>
            </TouchableOpacity>
          </View>
        </Container>
      </StyledScrollView>
    );
  }
  ```

```

```
