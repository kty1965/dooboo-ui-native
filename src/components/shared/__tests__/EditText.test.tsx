import 'react-native';

import * as React from 'react';

import {
  RenderResult,
  act,
  fireEvent,
  render,
} from '@testing-library/react-native';

import EditText from '../EditText';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

let props: any;
let component: React.ReactElement;
let testingLib: RenderResult;

describe('[EditText]', () => {
  let value = '';

  beforeEach(() => {
    props = {};
    component = <EditText {...props} />;
  });

  it('renders without crashing', () => {
    const rendered = renderer.create(component).toJSON();
    expect(rendered).toMatchSnapshot();
    expect(rendered).toBeTruthy();
  });

  describe('interactions', () => {
    const props = {
      testID: 'INPUT_TEST',
      testError: 'ERROR_TEST',
      onChangeText: (word: string): void => {
        value = word;
      },
    };
    beforeEach(() => {
      component = <EditText {...props} />;
      testingLib = render(component);
    });

    it('should set error message when no valid email has been written', async () => {
      const input = testingLib.getByTestId('INPUT_TEST');
      act(() => {
        fireEvent.changeText(input, 'input test');
      });
      expect(value).toEqual('input test');
      // expect(input.props).toEqual('input test');
    });
  });
});
