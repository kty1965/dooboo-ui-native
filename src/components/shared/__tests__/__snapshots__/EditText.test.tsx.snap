// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[EditText] renders without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
        "flexDirection": "column",
        "width": "100%",
      },
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#b9b9c4",
          "fontSize": 14,
          "fontWeight": "500",
        },
        null,
        undefined,
      ]
    }
  />
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    onBlur={[Function]}
    onFocus={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "color": "#2c374e",
          "fontSize": 15,
          "fontWeight": "500",
          "paddingBottom": 10,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 15,
        },
      ]
    }
    underlineColorAndroid="transparent"
  />
  <View
    style={
      Array [
        Object {
          "borderColor": "#eaeaf9",
          "borderStyle": "solid",
          "borderWidth": 0.6,
        },
        null,
        undefined,
      ]
    }
  />
</View>
`;
